engines/plumbers/3do.cpp

```diff
@@ -0,0 +1,703 @@
+		switch(scNo) {
+		case 4:
+		case 5:
+			_scenes[scPtr]._choices[0]._sceneName = "miketest/sc06";
+			break;
+		case 11:
+			_scenes[scPtr]._choices[0]._sceneName = "miketest/sc13";
+			break;
+		case 13:
+			_scenes[scPtr]._choices[0]._sceneName = "tuntest/dec/dec13";
+			break;
+		case 7:
+		case 8:
+		case 12:
+			_scenes[scPtr]._choices[0]._sceneName = Common::String::format("miketest/sc%02da", scNo);
+			break;
+		default:
+			_scenes[scPtr]._choices[0]._sceneName = Common::String::format("miketest/sc%02d", scNo + 1);
+			break;
+		}
+	}
+
+	makeMikeDecision(_scenes[scPtr], 3);
+	_scenes[scPtr]._sceneName = "miketest/sc07a";
+	_scenes[scPtr]._decisionBitmap = "DEC07";
+	_scenes[scPtr]._choices[0]._sceneName = "miketest/sc08";
+	_scenes[scPtr]._choices[0]._points = -10000;
+	_scenes[scPtr]._choices[1]._sceneName = "miketest/sc11";
+	_scenes[scPtr]._choices[1]._points = 10000;
+	_scenes[scPtr]._choices[2]._sceneName = "miketest/sc12";
+	_scenes[scPtr]._choices[2]._points = -20000;
+	scPtr++;
+
+	makeMikeDecision(_scenes[scPtr], 2);
+	_scenes[scPtr]._sceneName = "miketest/sc08a";
+	_scenes[scPtr]._decisionBitmap = "DEC08";
+	_scenes[scPtr]._choices[0]._sceneName = "miketest/sc09";
+	_scenes[scPtr]._choices[0]._points = 0;
+	_scenes[scPtr]._choices[1]._sceneName = "miketest/sc09";
+	_scenes[scPtr]._choices[1]._points = 10000;
+	scPtr++;
+
+	makeMikeDecision(_scenes[scPtr], 2);
+	_scenes[scPtr]._sceneName = "miketest/sc12a";
+	_scenes[scPtr]._decisionBitmap = "DEC12";
+	_scenes[scPtr]._choices[0]._sceneName = "miketest/sc07a";
+	_scenes[scPtr]._choices[0]._points = 0;
+	_scenes[scPtr]._choices[1]._sceneName = "restart";
+	_scenes[scPtr]._choices[1]._points = 0;
+	scPtr++;
+
+	for (uint scNo = 15; scNo <= 32; scNo++) {
+		// there is no sc19
+		if (scNo == 19)
+			continue;
+		Common::String imgScene = Common::String::format("sc%02d", scNo);
+		_scenes[scPtr]._bitmapNum = imgCounter[imgScene];
+		_scenes[scPtr]._startBitmap = firstImg[imgScene];
+		_scenes[scPtr]._sceneName = Common::String::format("tuntest/sc%02d", scNo);
+		_scenes[scPtr]._waveFilename = Common::String::format("sc%02d.aiff", scNo);
+		_scenes[scPtr]._style = Scene::STYLE_PC;
+		_scenes[scPtr]._decisionChoices = 1;
+		if (scNo == 32)
+			_scenes[scPtr]._choices[0]._sceneName = "end";
+		else if (scNo == 16 || scNo == 18 || scNo == 20 || scNo == 22 || scNo == 23 || scNo == 25
+		    || scNo == 26 || scNo == 28 || scNo == 29 || scNo == 30 || scNo == 31)
+			_scenes[scPtr]._choices[0]._sceneName = Common::String::format("tuntest/dec/dec%02d", scNo);
+		else
+			_scenes[scPtr]._choices[0]._sceneName = Common::String::format("tuntest/sc%02d", scNo + 1);
+		scPtr++;
+	}
+
+	file.seek(0x20290);
+
+	for (int i = 0; i < 26; i++) {
+		char buf[16];
+		file.read(buf, 16);
+		uint32 x = file.readUint32BE();
+		uint32 y = file.readUint32BE();
+		uint32 w = file.readUint32BE();
+		uint32 h = file.readUint32BE();
+		Common::String shortName = Common::String(buf).substr(0, 5);
```

Comment:
Ugh, this seems to be dangerous. Please write `Common::String(buf, 5)`. If bug is not null-terminated, you will have problem here and `String(const char *str, uint32 len)` is more secure and does exactly what you wrote with substr()

Reply:
Done